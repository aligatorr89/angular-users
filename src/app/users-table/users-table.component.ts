import { Component, OnInit, ViewChild, OnDestroy } from '@angular/core';
import { Subscription, Observable } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { MatSort } from '@angular/material/sort';
import { MatTableDataSource } from '@angular/material/table';
import { SelectionModel } from '@angular/cdk/collections';
import { IUser } from '../User';
import { UsersService } from '../angular-services/users.service';
import * as UsersAction from '../ngrx-store/user.actions';
import * as fromUsers from '../ngrx-store/user.reducer';

@Component({
  selector: 'app-users-table',
  templateUrl: './users-table.component.html',
  styleUrls: ['./users-table.component.css']
})
export class UsersTableComponent implements OnInit, OnDestroy {

  private usersSubscription$: Subscription;
  private users$: Observable<IUser[]>;

  constructor(private usersService: UsersService, private store: Store<IUser[]>) {
    console.log('constructor');
    // this.users$ = store.pipe(select(fromUsers.reducer));
  }

  displayedCols: string[] = ['select', 'id', 'name', 'email'];
  dataSource = new MatTableDataSource<IUser>();
  selection = new SelectionModel<IUser>(true, []);

  @ViewChild(MatSort, {static: true})
  sort: MatSort;

  ngOnInit() {
    this.getUsers();
    this.dataSource.sort = this.sort;
    // this.getUsersStore();
  }

  ngOnDestroy(): void {
    this.usersSubscription$.unsubscribe();
  }

  getUsers() {
    this.usersSubscription$ = this.usersService.getUsers().subscribe(users => this.dataSource.data = users);
  }

  getUsersStore() {
    console.log('usersStore here!');
    this.store.dispatch(UsersAction.getUsers({users: []}));
  }

  applyFilter(filterValue: string) {
    this.dataSource.filter = filterValue.trim().toLowerCase();
    this.sumId();
  }

  /** Whether the number of selected elements matches the total number of rows. */
  isAllSelected() {
    const numSelected = this.selection.selected.length;
    const numRows = this.dataSource.data.length;
    return numSelected === numRows;
  }

  /** Selects all rows if they are not all selected; otherwise clear selection. */
  masterToggle() {
    this.isAllSelected() ?
     this.selection.clear() :
     this.dataSource.data.forEach(row => this.selection.select(row));
  }

  /** The label for the checkbox on the passed row */
  checkboxLabel(row?: IUser): string {
    if (!row) {
      return `${this.isAllSelected() ? 'select' : 'deselect'} all`;
    }
    return `${this.selection.isSelected(row) ? 'deselect' : 'select'} row ${row.id + 1}`;
  }

  sumId() {
    return this.dataSource.filteredData.map(u => u.id).reduce((acc, value) => acc + value, 0);
  }



}
