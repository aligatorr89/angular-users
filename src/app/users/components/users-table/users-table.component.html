<div *ngIf="users">
  <mat-form-field>
    <input matInput (keyup)="applyFilter($event.target.value)" placeholder="Filter">
  </mat-form-field>
  <table mat-table [dataSource]="dataSource" matSort class="mat-elevation-z8">
    <ng-container matColumnDef="select">
      <th mat-header-cell *matHeaderCellDef>
        <mat-checkbox (change)="$event ? masterToggle() : null"
                      [checked]="selection.hasValue() && isAllSelected()"
                      [indeterminate]="selection.hasValue() && !isAllSelected()"
                      [aria-label]="checkboxLabel()">
        </mat-checkbox>
      </th>
      <td mat-cell *matCellDef="let row">
        <mat-checkbox (click)="$event.stopPropagation()"
                      (change)="$event ? selection.toggle(row) : null"
                      [checked]="selection.isSelected(row)"
                      [aria-label]="checkboxLabel(row)">
        </mat-checkbox>
      </td>
      <td mat-footer-cell *matFooterCellDef>&nbsp;</td>
    </ng-container>

    <ng-container matColumnDef="id">
      <th mat-header-cell *matHeaderCellDef mat-sort-header>Id</th>
      <td mat-cell *matCellDef="let element">{{ element.id }}</td>
      <td mat-footer-cell *matFooterCellDef>Total</td>
    </ng-container>

    <ng-container matColumnDef="name">
      <th mat-header-cell *matHeaderCellDef mat-sort-header>Name</th>
      <td mat-cell *matCellDef="let element">{{ element.name }}</td>
      <td mat-footer-cell *matFooterCellDef> {{ sumId() | currency }} </td>
    </ng-container>

    <ng-container matColumnDef="email">
      <th mat-header-cell *matHeaderCellDef mat-sort-header>Email</th>
      <td mat-cell *matCellDef="let element">{{ element.email }}</td>
      <td mat-footer-cell *matFooterCellDef>&nbsp;</td>
    </ng-container>

    <tr mat-header-row *matHeaderRowDef="displayedCols; sticky: true"></tr>
    <tr mat-row *matRowDef="let row; columns: displayedCols;"
      (click)="selection.toggle(row)">
    </tr>
    <tr mat-footer-row *matFooterRowDef="displayedCols; sticky: true"></tr>
  </table>
</div>
