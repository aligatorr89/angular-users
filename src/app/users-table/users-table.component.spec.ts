import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { AngularMaterialModule } from '../angular-material.module';
import { Store } from '@ngrx/store';
import { provideMockStore, MockStore } from '@ngrx/store/testing';

import { UsersTableComponent } from './users-table.component';
import { IUser, mockedUsers } from '../User';

describe('UsersTableComponent', () => {
  let component: UsersTableComponent;
  let fixture: ComponentFixture<UsersTableComponent>;
  let componentDom: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ UsersTableComponent ],
      imports: [
        HttpClientTestingModule,
        AngularMaterialModule
      ],
      providers: [
        provideMockStore()
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UsersTableComponent);
    component = fixture.componentInstance;
    componentDom = fixture.debugElement.nativeElement;

    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match snapshot', () => {
    // expect(componentDom).toMatchSnapshot()
    expect(1).toBe(1);
  })
});
