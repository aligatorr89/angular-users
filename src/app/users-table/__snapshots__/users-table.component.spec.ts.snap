// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTableComponent should match snapshot 1`] = `
<div
  id="root1"
  ng-version="8.0.2"
>
  <div>
    <mat-form-field
      class="mat-form-field ng-tns-c4-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder"
    >
      <div
        class="mat-form-field-wrapper"
      >
        <div
          class="mat-form-field-flex"
        >
          
          
          <div
            class="mat-form-field-infix"
          >
            <input
              aria-invalid="false"
              aria-required="false"
              class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
              id="mat-input-1"
              matinput=""
              ng-reflect-placeholder="Filter"
              placeholder="Filter"
            />
            <span
              class="mat-form-field-label-wrapper"
            >
              
              <label
                aria-owns="mat-input-1"
                class="mat-form-field-label ng-tns-c4-0 mat-empty mat-form-field-empty ng-star-inserted"
                for="mat-input-1"
                id="mat-form-field-label-3"
                ng-reflect-disabled="true"
                ng-reflect-ng-switch="false"
              >
                
                
                Filter
                
                
              </label>
            </span>
          </div>
          
        </div>
        
        <div
          class="mat-form-field-underline ng-tns-c4-0 ng-star-inserted"
        >
          <span
            class="mat-form-field-ripple"
          />
        </div>
        <div
          class="mat-form-field-subscript-wrapper"
          ng-reflect-ng-switch="hint"
        >
          
          
          <div
            class="mat-form-field-hint-wrapper ng-tns-c4-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
            style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
          >
            
            <div
              class="mat-form-field-hint-spacer"
            />
          </div>
        </div>
      </div>
    </mat-form-field>
    <table
      class="mat-elevation-z8 mat-table"
      mat-table=""
      matsort=""
      ng-reflect-data-source="[object Object]"
      role="grid"
    >
      <thead
        role="rowgroup"
      >
        
        <tr
          class="mat-header-row ng-star-inserted"
          mat-header-row=""
          role="row"
        >
          
          <th
            class="mat-header-cell cdk-column-select mat-column-select mat-table-sticky ng-star-inserted"
            mat-header-cell=""
            role="columnheader"
            style="top: 0px; position: sticky; z-index: 100;"
          >
            <mat-checkbox
              class="mat-checkbox mat-accent"
              id="mat-checkbox-2"
              ng-reflect-aria-label="select all"
              ng-reflect-checked="false"
              ng-reflect-indeterminate="false"
            >
              <label
                class="mat-checkbox-layout"
                for="mat-checkbox-2-input"
              >
                <div
                  class="mat-checkbox-inner-container mat-checkbox-inner-container-no-side-margin"
                >
                  <input
                    aria-checked="false"
                    aria-label="select all"
                    class="mat-checkbox-input cdk-visually-hidden"
                    id="mat-checkbox-2-input"
                    tabindex="0"
                    type="checkbox"
                  />
                  <div
                    class="mat-checkbox-ripple mat-ripple"
                    matripple=""
                    ng-reflect-animation="[object Object]"
                    ng-reflect-centered="true"
                    ng-reflect-disabled="false"
                    ng-reflect-radius="20"
                    ng-reflect-trigger="[object HTMLLabelElement]"
                  >
                    <div
                      class="mat-ripple-element mat-checkbox-persistent-ripple"
                    />
                  </div>
                  <div
                    class="mat-checkbox-frame"
                  />
                  <div
                    class="mat-checkbox-background"
                  >
                    <svg
                      class="mat-checkbox-checkmark"
                      focusable="false"
                      version="1.1"
                      viewBox="0 0 24 24"
                      xml:space="preserve"
                    >
                      <path
                        class="mat-checkbox-checkmark-path"
                        d="M4.1,12.7 9,17.6 20.3,6.3"
                        fill="none"
                        stroke="white"
                      />
                    </svg>
                    <div
                      class="mat-checkbox-mixedmark"
                    />
                  </div>
                </div>
                <span
                  class="mat-checkbox-label"
                >
                  <span
                    style="display:none"
                  >
                    Â 
                  </span>
                </span>
              </label>
            </mat-checkbox>
          </th>
          <th
            class="mat-header-cell ng-tns-c7-1 cdk-column-id mat-column-id mat-table-sticky ng-star-inserted"
            mat-header-cell=""
            mat-sort-header=""
            ng-reflect-id=""
            role="columnheader"
            style="top: 0px; position: sticky; z-index: 100;"
          >
            <div
              class="mat-sort-header-container"
            >
              <button
                aria-label="Change sorting for id"
                class="mat-sort-header-button"
                type="button"
              >
                Id
              </button>
              
              <div
                class="mat-sort-header-arrow ng-trigger ng-trigger-arrowPosition ng-tns-c7-1 ng-star-inserted"
                style="transform:translateY(25%);opacity:0;0:transform;1:opacity;"
              >
                <div
                  class="mat-sort-header-stem"
                />
                <div
                  class="mat-sort-header-indicator ng-trigger ng-trigger-indicator"
                  style="transform:translateY(0px);0:transform;"
                >
                  <div
                    class="mat-sort-header-pointer-left ng-trigger ng-trigger-leftPointer"
                    style="transform:rotate(-45deg);0:transform;"
                  />
                  <div
                    class="mat-sort-header-pointer-right ng-trigger ng-trigger-rightPointer"
                    style="transform:rotate(45deg);0:transform;"
                  />
                  <div
                    class="mat-sort-header-pointer-middle"
                  />
                </div>
              </div>
            </div>
          </th>
          <th
            class="mat-header-cell ng-tns-c7-2 cdk-column-name mat-column-name mat-table-sticky ng-star-inserted"
            mat-header-cell=""
            mat-sort-header=""
            ng-reflect-id=""
            role="columnheader"
            style="top: 0px; position: sticky; z-index: 100;"
          >
            <div
              class="mat-sort-header-container"
            >
              <button
                aria-label="Change sorting for name"
                class="mat-sort-header-button"
                type="button"
              >
                Name
              </button>
              
              <div
                class="mat-sort-header-arrow ng-trigger ng-trigger-arrowPosition ng-tns-c7-2 ng-star-inserted"
                style="transform:translateY(25%);opacity:0;0:transform;1:opacity;"
              >
                <div
                  class="mat-sort-header-stem"
                />
                <div
                  class="mat-sort-header-indicator ng-trigger ng-trigger-indicator"
                  style="transform:translateY(0px);0:transform;"
                >
                  <div
                    class="mat-sort-header-pointer-left ng-trigger ng-trigger-leftPointer"
                    style="transform:rotate(-45deg);0:transform;"
                  />
                  <div
                    class="mat-sort-header-pointer-right ng-trigger ng-trigger-rightPointer"
                    style="transform:rotate(45deg);0:transform;"
                  />
                  <div
                    class="mat-sort-header-pointer-middle"
                  />
                </div>
              </div>
            </div>
          </th>
          <th
            class="mat-header-cell ng-tns-c7-3 cdk-column-email mat-column-email mat-table-sticky ng-star-inserted"
            mat-header-cell=""
            mat-sort-header=""
            ng-reflect-id=""
            role="columnheader"
            style="top: 0px; position: sticky; z-index: 100;"
          >
            <div
              class="mat-sort-header-container"
            >
              <button
                aria-label="Change sorting for email"
                class="mat-sort-header-button"
                type="button"
              >
                Email
              </button>
              
              <div
                class="mat-sort-header-arrow ng-trigger ng-trigger-arrowPosition ng-tns-c7-3 ng-star-inserted"
                style="transform:translateY(25%);opacity:0;0:transform;1:opacity;"
              >
                <div
                  class="mat-sort-header-stem"
                />
                <div
                  class="mat-sort-header-indicator ng-trigger ng-trigger-indicator"
                  style="transform:translateY(0px);0:transform;"
                >
                  <div
                    class="mat-sort-header-pointer-left ng-trigger ng-trigger-leftPointer"
                    style="transform:rotate(-45deg);0:transform;"
                  />
                  <div
                    class="mat-sort-header-pointer-right ng-trigger ng-trigger-rightPointer"
                    style="transform:rotate(45deg);0:transform;"
                  />
                  <div
                    class="mat-sort-header-pointer-middle"
                  />
                </div>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        
      </tbody>
      <tfoot
        class="mat-table-sticky"
        role="rowgroup"
        style="bottom: 0px; position: sticky; z-index: 10;"
      >
        
        <tr
          class="mat-footer-row ng-star-inserted"
          mat-footer-row=""
          role="row"
        >
          
          <td
            class="mat-footer-cell cdk-column-select mat-column-select mat-table-sticky ng-star-inserted"
            mat-footer-cell=""
            role="gridcell"
            style="bottom: 0px; position: sticky; z-index: 10;"
          >
            Â 
          </td>
          <td
            class="mat-footer-cell cdk-column-id mat-column-id mat-table-sticky ng-star-inserted"
            mat-footer-cell=""
            role="gridcell"
            style="bottom: 0px; position: sticky; z-index: 10;"
          >
            Total
          </td>
          <td
            class="mat-footer-cell cdk-column-name mat-column-name mat-table-sticky ng-star-inserted"
            mat-footer-cell=""
            role="gridcell"
            style="bottom: 0px; position: sticky; z-index: 10;"
          >
             $0.00 
          </td>
          <td
            class="mat-footer-cell cdk-column-email mat-column-email mat-table-sticky ng-star-inserted"
            mat-footer-cell=""
            role="gridcell"
            style="bottom: 0px; position: sticky; z-index: 10;"
          >
            Â 
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
